File: go-budget-api/internal/services/report_service.go
package services

import (
	"budget-api/internal/db"
	"budget-api/internal/entities"
	"budget-api/internal/models"
	"time"
)

func GetReport(credentialId string, category *string, from *time.Time, to *time.Time) (*models.BudgetReport, error) {
	query := db.Context.Model(&entities.Transaction{}).Where("credential_id = ?", credentialId)

	if category != nil {
		query = query.Where("category = ?", *category)
	}

	if from != nil {
		query = query.Where("timestamp >= ?", from.UTC())
	}

	if to != nil {
		query = query.Where("timestamp <= ?", to.UTC())
	}

	var transactionValues []float64
	if err := query.Select("value").Find(&transactionValues).Error; err != nil {
		return nil, err
	}

	return GenerateReport(&transactionValues), nil
}

func GenerateReport(values *[]float64) *models.BudgetReport {
	report := models.BudgetReport{}

	for _, value := range *values {
		isIncome := value > 0
		isExpense := value < 0

		report.TransactionsSum += value
		report.TransactionsCount++

		if isIncome {
			report.IncomesSum += value
			report.IncomesCount++
		}

		if isExpense {
			report.ExpensesSum += value
			report.ExpensesCount++
		}
	}

	return &report
}

File: node-budget-api/src/services/report-service.js
const { Transaction } = require("../models");
const {
  roundToTwoDecimals,
  getTransactionsWhereClause,
} = require("../utils/helpers");

const initialReport = {
  transactions_sum: 0,
  expenses_sum: 0,
  incomes_sum: 0,
  transactions_count: 0,
  expenses_count: 0,
  incomes_count: 0,
};

const mapReport = (values) =>
  values.reduce((report, value) => {
    const isExpense = value < 0;
    const isIncome = value > 0;

    return {
      transactions_sum: roundToTwoDecimals(report.transactions_sum + value),
      expenses_sum: isExpense
        ? roundToTwoDecimals(report.expenses_sum + value)
        : report.expenses_sum,
      incomes_sum: isIncome
        ? roundToTwoDecimals(report.incomes_sum + value)
        : report.incomes_sum,
      transactions_count: report.transactions_count + 1,
      expenses_count: isExpense
        ? report.expenses_count + 1
        : report.expenses_count,
      incomes_count: isIncome ? report.incomes_count + 1 : report.incomes_count,
    };
  }, initialReport);

const generateReport = async (credentialId, category, from, to) => {
  const where = getTransactionsWhereClause(credentialId, category, from, to);
  const transactions = await Transaction.findAll({
    where,
    attributes: ["value"],
  });

  const values = transactions.map((transaction) => Number(transaction.value));

  return mapReport(values);
};

module.exports = { generateReport };
